### PR TITLE
Fix v3 behaviour when input or output token is ETH

### PR DESCRIPTION
**Pull request by nimcon**
Fix for Uniswap v3 behaviour when ETH is the input or output token, resolving issues #154 and #155. There has also been some general tidying-up of parts of the code. Basic tests for all v3 scenarios have been run, but it would be good to check edge cases and verify that v1 and v2 still work as expected.

### CODE REVIEW

File: uniswap/uniswap.py [hunk 1]
```diff
@@ -537,11 +584,45 @@ def _token_to_eth_swap_input(
...
+            swap_selector = function_signature_to_4byte_selector(f"exactInputSingle({swap_arg_types_str})")
+
+            swap_args = [input_token, output_token, fee, ETH_ADDRESS, self._deadline(), qty, min_tokens_bought, sqrtPriceLimitX96]
+            swap_args_encoded = encode_abi(swap_arg_types, swap_args)
+
+            # Prepare unwrap data
+            unwrap_arg_types = ["uint256", "address"]
+            unwrap_arg_types_str = f"{','.join(unwrap_arg_types)}"
+
+            unwrap_selector = function_signature_to_4byte_selector(f"unwrapWETH9({unwrap_arg_types_str})")
```
Comment (romanbsd) [inline uniswap/uniswap.py:610]: You can use contract.encodeABI("unwrapWETH9", [eth_amount, recipient]) instead of doing it manually, this way you don't have to specify the arg types, since the contract is already aware of them.

**Comment by romanbsd** [comment]
Cool! I didn't get to do it, and it's pretty much what I had in mind!

File: uniswap/uniswap.py [hunk 2]
```diff
@@ -376,28 +395,27 @@ def make_trade(
         if slippage is None:
             slippage = self.default_slippage
 
+        if input_token == output_token:
+            raise ValueError
```
Comment (romanbsd) [inline uniswap/uniswap.py:399]: maybe `assert input_token != output_token` is better?

File: uniswap/uniswap.py [hunk 3]
```diff
@@ -37,6 +37,9 @@
     ETH_ADDRESS,
 )
 
+from eth_abi import encode_abi
+from eth_utils import function_signature_to_4byte_selector
```
Comment (ErikBjare) [inline uniswap/uniswap.py:41]:
```suggestion
```

No longer used.

**Review by ErikBjare — COMMENTED** [review]
Sorry for the delay. Ran tests locally and works well!

A couple of unrelated tests failed, but that's my bad, and they're not testing anything important.
